flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartition.java

```diff
@@ -375,4 +397,29 @@ void onConsumedSubpartition(int subpartitionIndex) {
 	private void checkInProduceState() throws IllegalStateException {
 		checkState(!isFinished, "Partition already finished.");
 	}
+
+	/**
+	 * Check whether all subpartitions' backlogs are less than the limitation of max backlogs, and make this partition
+	 * available again if yes.
+	 */
+	public void notifyDecreaseBacklog(int buffersInBacklog) {
+		if (buffersInBacklog == maxBuffersPerChannel) {
+			if (--unavailableSubpartitionsCount == 0) {
+				CompletableFuture<?> toNotify = availabilityHelper.getUnavailableToResetAvailable();
+				toNotify.complete(null);
```

Comment:
this is being executed under the `buffers` lock. In the past we had quite a bit of problem of executing callbacks under a lock, leading to deadlocks. However since we introduced mailbox model, it might be less risky to do so, as I think the only thing this call back can do is enqueue something into the mailbox/resume default action processing? What do you think @wsry ?

Reply:
I think you are right, the callback only enqueues a resume mail so there should be no deadlock issue.